pkg/reconciler/apis/apibinding/apibinding_reconcile_permissionclaim.go

```diff
@@ -0,0 +1,302 @@
+	key, label, err := permissionclaims.PermissionClaimToLabel(pc)
+	if err != nil {
+		return permissionClaimHelper{}, gvr, err
+	}
+
+	return permissionClaimHelper{
+		claim:  pc,
+		label:  label,
+		key:    key,
+		lister: lister,
+	}, gvr, nil
+}
+
+// reconcilePermissionClaims will determine the resources that need to be labeled for access by a permission claim.
+// It will determine what permssions need to be added, what permissions need to be removed.
+// It will also update the status if it finds an invalid permission claim.
+// Permission claims are considered invalid when the identity hashes are mismatched, and when there is no dynamic informer
+// for the group resource.
+func (c *controller) reconcilePermissionClaims(ctx context.Context, apiBinding *apisv1alpha1.APIBinding) error {
+	identityMismatch := false
+	lc := logicalcluster.From(apiBinding)
+
+	// Get all the bound resources in the locicalcluster
+	bindings, err := c.listAPIBindings(lc)
+	if err != nil {
+		return err
+	}
+	grsToBoundResource := map[apisv1alpha1.GroupResource]apisv1alpha1.BoundAPIResource{}
+	for _, binding := range bindings {
+		for _, resource := range binding.Status.BoundResources {
+			grsToBoundResource[apisv1alpha1.GroupResource{Group: resource.Group, Resource: resource.Resource}] = resource
+		}
+	}
+
+	invalidClaims := []apisv1alpha1.PermissionClaim{}
+	// get the added permission claims if added, add are just requeue the object.
+	addedPermissionClaims := mapGVRToClaim{}
+	for _, acceptedPC := range apiBinding.Spec.AcceptedPermissionClaims {
+		found := false
+		for _, observedPC := range apiBinding.Status.ObservedAcceptedPermissionClaims {
+			if acceptedPC.Equal(observedPC) {
+				found = true
+				break
+			}
+		}
+		if !found {
+			if acceptedPC.IdentityHash != "" {
+				if grsToBoundResource[acceptedPC.GroupResource].Schema.IdentityHash != acceptedPC.IdentityHash {
+					identityMismatch = true
+					invalidClaims = append(invalidClaims, acceptedPC)
+					continue
+				}
+			}
+			claim, gvr, err := c.createPermissionClaimHelper(acceptedPC)
+			if err != nil {
+				return err
+			}
+			addedPermissionClaims[gvr] = append(addedPermissionClaims[gvr], claim)
+		}
+	}
+
+	removedPermissionClaims := mapGVRToClaim{}
+	for _, observedPC := range apiBinding.Status.ObservedAcceptedPermissionClaims {
+		found := false
+		for _, acceptedPC := range apiBinding.Spec.AcceptedPermissionClaims {
+			if observedPC.Equal(acceptedPC) {
+				found = true
+				break
+			}
+		}
+		if !found {
+			claim, gvr, err := c.createPermissionClaimHelper(observedPC)
+			if err != nil {
+				return err
+			}
+			claims := removedPermissionClaims[gvr]
+			if claims == nil {
+				claims = []permissionClaimHelper{}
+			}
+			claims = append(claims, claim)
+			removedPermissionClaims[gvr] = claims
+		}
+	}
+
+	klog.V(2).Infof("permission controller adding permission claims: %v, removing: %v and found invalid claims: %s", addedPermissionClaims, removedPermissionClaims, invalidClaims)
```

Comment:
(not a blocker) this is too verbose. If nothing change, this must be `klog.V(4)`, see https://github.com/kcp-dev/kcp/blob/main/CONTRIBUTING.md#coding-guidelines--conventions.

Reply:
updated